src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs

```diff
@@ -0,0 +1,490 @@
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            FunctionStartedEvent startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                object convertedInput = input;
+                if (input != null)
+                {
+                    HttpRequestMessage request = input as HttpRequestMessage;
+                    if (request != null)
+                    {
+                        // TODO: Handle other content types? (E.g. byte[])
+                        if (request.Content != null && request.Content.Headers.ContentLength > 0)
+                        {
+                            convertedInput = ((HttpRequestMessage)input).Content.ReadAsStringAsync().Result;
+                        }
+                    }
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function started (Id={0})", invocationId));
+
+                string functionInstanceOutputPath = Path.Combine(Path.GetTempPath(), "Functions", "Binding",
+                    invocationId);
+
+                Dictionary<string, string> bindingData = GetBindingData(convertedInput, binder);
+                bindingData["InvocationId"] = invocationId;
+
+                await
+                    ProcessInputBindingsAsync(convertedInput, functionInstanceOutputPath, binder, bindingData,
+                        _environmentVariables);
+
+                InitializeEnvironmentVariables(_environmentVariables, functionInstanceOutputPath, input,
+                    functionExecutionContext);
+
+                InvokePowerShellScript();
+
+                await
+                    ProcessOutputBindingsAsync(functionInstanceOutputPath, _outputBindings, input, binder, bindingData);
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Success, Id={0})", invocationId));
+            }
+            catch (Exception exception)
+            {
+                startedEvent.Success = false;
+                RuntimeException runtimeException = exception as RuntimeException;
+                if (runtimeException != null)
+                {
+                    TraceWriter.Verbose(string.Format(
+                        "Function runtime exception: {0}: {1}{2}{3}",
+                        runtimeException.ErrorRecord.InvocationInfo.InvocationName,
+                        runtimeException.Message,
+                        Environment.NewLine,
+                        runtimeException.ErrorRecord.InvocationInfo.PositionMessage));
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Failure, Id={0})", invocationId));
+                throw;
+            }
+            finally
+            {
+                _metrics.EndEvent(startedEvent);
+                DeleteProcessEnvironmentVariables(_environmentVariables);
```

Comment:
Hmmm, this might be a problem. In the ScriptInvoker case, these environment variables are being injected into a child process, so they don't pollute the parent process, and there is not possibility of collisions.

In this case, if you have N instances of a function running concurrently, won't they overwrite each others state? I think we need a way to inject these into the PowerShell host so they aren't parent env vars.